tremor-connectors-azure/src/rest.rs

```diff
@@ -0,0 +1,222 @@
+use tremor_value::literal;
+use tremor_value::{StaticNode, Value};
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) struct RequestId(u64);
+
+impl RequestId {
+    pub(crate) fn new(id: u64) -> Self {
+        Self(id)
+    }
+    pub(crate) fn get(self) -> u64 {
+        self.0
+    }
+}
+
+impl std::fmt::Display for RequestId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+pub(crate) struct RequestMeta {
+    method: String,
+    url: String,
+    headers: Headers,
+}
+
+impl RequestMeta {
+    pub fn default() -> Self {
+        Self {
+            method: String::new(),
+            url: String::new(),
+            headers: Headers::new(),
+        }
+    }
+}
+
+impl From<RequestMeta> for Value<'static> {
+    fn from(meta: RequestMeta) -> Value<'static> {
+        let mut headers =
+            tremor_value::Object::with_capacity_and_hasher(5, ObjectHasher::default());
+        for (key, value) in meta.headers.iter() {
+            headers.insert(
+                Cow::owned(key.as_str().to_string()),
+                Value::String(Cow::owned(value.as_str().to_string())),
+            );
+        }
+
+        literal!(
+            {
+                "method": meta.method,
+                "url": meta.url,
+                "headers": Value::Object(Box::new(headers)),
+            }
+        )
+    }
+}
+
+pub(crate) struct ResponseMeta {
+    status: u16,
+    headers: Headers,
+    pub(crate) data: Value<'static>,
+    pub(crate) content_length: usize,
+}
+
+impl From<ResponseMeta> for Value<'static> {
+    fn from(meta: ResponseMeta) -> Value<'static> {
+        let mut headers =
+            tremor_value::Object::with_capacity_and_hasher(5, ObjectHasher::default());
+        for (key, value) in meta.headers.iter() {
+            headers.insert(
+                Cow::owned(key.as_str().to_string()),
+                Value::String(Cow::owned(value.as_str().to_string())),
+            );
+        }
+
+        literal!(
+            {
+                "status": meta.status,
+                "headers": Value::Object(Box::new(headers)),
+                "content_length": meta.content_length,
+                "data": meta.data
+            }
+        )
+    }
+}
+
+impl ResponseMeta {
+    pub fn default() -> Self {
+        Self {
+            status: 0,
+            headers: Headers::new(),
+            data: Value::Static(StaticNode::Null),
+            content_length: 0,
+        }
+    }
+}
+
+#[allow(clippy::unnecessary_wraps)]
+pub(crate) fn extract_request_meta(request: &Request) -> Result<RequestMeta, Error> {
+    let mut meta = RequestMeta::default();
+    meta.method = request.method().to_string();
+    meta.url = request.url().to_string();
+    meta.headers = request.headers().clone();
+    Ok(meta)
+}
+
+pub(crate) async fn extract_response_meta(response: Response) -> Result<ResponseMeta, Error> {
```

Comment:
This seems to be mixing data and metadata. I think we wan't the body data in the event data section not the metadata. that also avoids the casting to static and avoids a bunch of cloning

Reply:
I didn't deviate much from HTTP client structurally so its no different. The azure-core crate basically builds on reqwest and i'd like to see if we can drop straight to reqwest. That way, perhaps we can drop hyper client from the http connector too. If that works, possibly worth revisiting how we do req/resp metadata.

Punting for now.

Reply:
you said that we always get 204's back so we don't need the body at all,  I guess it's fine then. What I want to avoid is that we release the body in the metadata as that would break all conventions and expectations we established in all the other connectors.

Reply:
just looked it up, yap it's the one that [never gets use](https://github.com/tremor-rs/tremor-runtime/pull/2583/files/4e3a5785b860270c65afd3a1d34a8b3ca8df551e#diff-b35c45032121e1ecb6dcd1eaacadd2ad13a968fd4de464c0d5905847ac357cb9R211)

perhaps just exclude the body from this as it has no effect and we avoid accidentially pulling it into the response later when we forgot about it?

Reply:
I stand corrected, it is used and pulled apart. What confused me here is the naming as it doesn't align with the use, it is used as a mix of respose data and response metadata and get's used in both places :D